test/e2e/networking.go

```diff
@@ -101,155 +100,227 @@ var _ = Describe("Networking", func() {
+			passed = true
+			return true, nil
+		case string(body) == "running":
+			Logf("test still running")
+		case string(body) == "fail":
+			if body, err = getDetails(); err != nil {
+				Failf("Failed on attempt. Cleaning up. Error reading details: %v", err)
+				return false, err
+			} else {
+				Failf("Failed on attempt. Cleaning up. Details:\n%s", string(body))
+				return false, nil
 			}
-			Logf("Only one ready node is detected. The test has limited scope in such setting. " +
-				"Rerun it with at least two nodes to get complete coverage.")
+		case strings.Contains(string(body), "no endpoints available"):
+			Logf("Attempt: waiting on service/endpoints")
+		default:
+			Logf("Unexpected response:\n%s", body)
+		}
+		return false, nil
+	}))
+
+	if !passed {
+		if body, err := getDetails(); err != nil {
+			Failf("Timed out. Cleaning up. Major error : Couldn't read service details: %v", err)
+		} else {
+			Failf("Timed out. Cleaning up. Service details :\n%s", string(body))
 		}
+	}
+	Logf("Sending final result to channel")
+	ch <- 0
+	Logf("RETURN")
+	return passed
+}
+
+//RunNetTest Creates a single pod on each host which serves
+//on a unique port in the cluster.  It then binds a service to
+//that port, so that there are "n" nodes to balance traffic to -
+//finally, each node reaches out to ping every other node in
+//the cluster on the given port.
+//The more ports you give, the more services will be spun up,
+//i.e. one service per port.
+//To test basic pod networking, send a single port.
+//To soak test the services, we can send a range (i.e. 8000-9000).
+func RunNetTest(doneTimeout Done, f *Framework, ports []int, nettestVersion string) {
+	defer GinkgoRecover()
+	defer close(doneTimeout)
+	//By("Creating a webserver (pending) pod on each node")
+
+	nodes, err := f.Client.Nodes().List(labels.Everything(), fields.Everything())
+	if err != nil {
+		Failf("Failed to list nodes: %v", err)
+	}
+	// previous tests may have cause failures of some nodes. Let's skip
+	// 'Not Ready' nodes, just in case (there is no need to fail the test).
+	filterNodes(nodes, func(node api.Node) bool {
+		return isNodeReadySetAsExpected(&node, true)
+	})
 
-		podNames := LaunchNetTestPodPerNode(f, nodes, svcname, "1.6")
+	if len(nodes.Items) == 0 {
+		Failf("No Ready nodes found.")
+	}
+	if len(nodes.Items) == 1 {
+		// in general, the test requires two nodes. But for local development, often a one node cluster
+		// is created, for simplicity and speed. (see issue #10012). We permit one-node test
+		// only in some cases
+		if !providerIs("local") {
+			Failf(fmt.Sprintf("The test requires two Ready nodes on %s, but found just one.", testContext.Provider))
+		}
+		Logf("Only one ready node is detected. The test has limited scope in such setting. " +
+			"Rerun it with at least two nodes to get complete coverage.")
+	}
+	sem := make(chan int, len(ports))
+	for ii := range ports {
+		//required to copy the var when we're creating it.
+		i := ports[ii]
```

Comment:
if you pass this as an argument to your goroutine below it shouldn't be necessary to make a copy here.